src/LinuxTracing/LinuxTracingUtils.cpp

```diff
@@ -213,4 +222,121 @@ bool SetMaxOpenFilesSoftLimit(uint64_t soft_limit) {
   return true;
 }
 
+// Check for the existence of a file mapping containing the absolute address of the function. We
+// have to consider the possibility that the module be mapped multiple times, and hence that the
+// function have multiple absolute addresses.
+//
+// Note: A more naive solution would be to look for a map containing the file offset for the
+// function, hence not involving absolute addresses and modules at all. For misaligned PEs, this can
+// cause false negatives, because a function can be mapped twice, in a file mapping and in an
+// anonymous (executable) mapping, but with the actual absolute address of the function being in the
+// anonymous (executable) mapping.
+//
+// Example: Consider a PE with one section, the .text section, at offset in the file 0x400 and
```

Comment:
> Would it be sufficient to disregard the non executable segments to remove the ambiguity described below?

Conceptually, unfortunately no. In the example I show the file mapping corresponding to the headers, followed by the file mapping corresponding to the section containing the function. However, there could be a case where the PE has two consecutive executable sections, and the first one happens to be aligned "correctly" (such that a file mapping is possible) by chance: we could have the same problem, with the first map also containing part of the second section, and for a function in the second section the problem would be the same, but this time the first map is also executable.

In other words: there could be a pair of mappings like in the example below, but with the file mapping also corresponding to an executable section and hence executable.

I hope this makes sense.

Reply:
> Also it might be helpful to sort out the non-executable segements for another reason: The code below looks as if it would not complain if it finds a function only in a non-executable segment belonging to a file mapping.

This is true and I would say this is intended here. Basically here we are considering valid cases that are "interesting" in the context of uprobes: module not yet loaded at all (not supported by Orbit, yet?, but supported by uprobes), and part of the module loaded with an anonymous map (but uprobes needs a file mapping).

In this context, a function not in an executable section is not a valid case and is out of scope: it means that something is wrong at a different level: either we made some mistake somewhere in computing where the function would be in memory, or the module is not fully mapped and the function will not get any hit from any kind of dynamic instrumentation simply because it won't be executed. If we really want to check for this case, we need to do it somewhere else, in a different context.

I hope this also makes sense.

Reply:
> and part of the module loaded with an anonymous map (but uprobes needs a file mapping).

~~That's the case I had in mind. Let's say there is a function at file offset 0x0500 in the example below. `FunctionIsInFileMapping` would return true since the function is "accidentally" in the non-executable file mapping (but gets executed from the executable anonymous mapping). So the function is inaccessible to uprobes but we don't generate a warning.~~

edit: And now I understood it: The `function_absolute_address` would be in the anonymous mapping and so `FunctionIsInFileMapping` would return false and that is exactly the case you want to exclude here.

Thanks for the explanations - makes sense now!

Reply:
Yep, that's exactly it.